feat(4646/4836): window period calculation

wiedld opened the pull request:
Part of #4646 
Part of #4836 


## Central questions being solved:
Which inputs to consider when calculating the windowPeriod? Which ones override others?
* Input from timeRange picker:
   * “Past 15m” etc have a preset timeRange.
   * “Custom date” gets a calculated timeRange.
* Where in the flux query is windowPeriod used?
   * On which range?
   * Does this range use the timeRange from the outer UI? or it's own?
* Does the user declare `v = {windowPeriod: <number>}` within the query itself?
* screenshot showing 3 sources of inputs, to resolve:
         <img width="1079" alt="Screen Shot 2022-06-28 at 11 49 07 PM" src="https://user-images.githubusercontent.com/10232835/176371374-e3344f19-ebb2-43d8-a15b-e8e70720d064.png">

## Issues pertaining to `v.windowPeriod`:
* not detecting the need to inject an extern
* not correctly calculating
* not scoping
* doesn't handle aliasing
* Very well detailed in the issue tickets linked above.
* Test fixtures are all in a single ` 5,125` line file. So the actual lines of code are less.

## Fix:
* handle all of that^^
* make test cases to show current contract
* is not used anywhere in the UI, during prod. (This will be the next PR.)


## Checklist

Authors and Reviewer(s), please verify the following:

- [x] A PR description, regardless of the triviality of this change, that communicates the value of this PR
- [x] [Well-formatted conventional commit messages](https://www.conventionalcommits.org/en/v1.0.0/) that provide context into the change
- [x] Documentation updated or issue created (provide link to issue/PR)

wiedld commented on src/shared/utils/duration/timeRange.ts:
```diff
@@ -0,0 +1,75 @@
+import {
```
This file is all legacy code.

wiedld commented on src/shared/utils/duration/utils.ts:
```diff
@@ -120,65 +112,3 @@ export const areDurationsEqual = (a: string, b: string): boolean => {
   }
 }
-
-export const timeRangeToDuration = (timeRange: TimeRange): string => {
```
Block of legacy code. Moved into it's own file.

wiedld commented:
> Super confused, I walked away from our meeting last week thinking that the status quo was fine. Am I missing something? Looks like this addresses problems that weren't brought up previously

As of yesterday, that is no longer the case. I was given the greenlight to move this forward.

kristinarobinson commented on src/shared/constants/timeRanges.ts:
```diff
@@ -135,3 +135,6 @@ export const SELECTABLE_USAGE_TIME_RANGES: SelectableDurationTimeRange[] = [
 export const DEFAULT_TIME_RANGE = pastHourTimeRange
+
+export const DESIRED_POINTS_PER_GRAPH = 360
+export const FALLBACK_WINDOW_PERIOD = 15000
```
I thought the default for windowPeriod was 10s, not 15s?

wiedld replied:
To be clear: this is the fallback if it doesn't recognize the timeRange used. Or something else fails. Most of the `SELECTABLE_TIME_RANGES` uses 10s as the windowPeriod.

wiedld commented on src/shared/constants/timeRanges.ts:
```diff
@@ -135,3 +135,6 @@ export const SELECTABLE_USAGE_TIME_RANGES: SelectableDurationTimeRange[] = [
 export const DEFAULT_USAGE_TIME_RANGE = pastDayTimeRange
 export const DEFAULT_TIME_RANGE = pastHourTimeRange
+
+export const DESIRED_POINTS_PER_GRAPH = 360
```
Existing value, defined in 2 places in our code.
I simply moved it into a single location.

wiedld commented on src/shared/utils/windowPeriod/fromAST.ts:
```diff
@@ -0,0 +1,117 @@
+  } = getWindowPeriodVariableAssignment(ast, outerContext)
+
+  if (!windowVariableAssignmentNode) {
+    throw new Error('windowPeriod not found in neither query nor outer scope')
```
There should always be a windowPeriod provided somewhere in the scope. Since we provide the timeRange from the  "outer UI scope" (e.g. the DateRange picker). If the dev forgets to pass in this timeRange information --> then it will fail. Forcing the dev to recognize the error prior to release.

wiedld replied:
Since our UI always has a sense of timeRange, then the contract is that a timeRange is always provided from a UI variable. From the DatePicker, with the ability to override in the flux query itself. Then the order of priority is:
1. if user provides a definition of `v.windowPeriod`, then take that
2. if not, then we calculate it. Based on the scoped order:
3. use the timeRange provided in the flux query, at the scoped point-of-use of the windowPeriod

wiedld commented on src/shared/utils/windowPeriod/fromDuration.ts:
```diff
@@ -0,0 +1,27 @@
+    tr => tr.seconds * 1000 === rangeDuration
+  )
+  if (foundDuration) {
+    return foundDuration.windowPeriod
```
we set the timeRange that we want, per the DatePicker, in `SELECTABLE_TIME_RANGES`. It may be override by the user in the internal flux query (which the visitor then finds). But we do want the the default window period is constant.

How does this method `convertMillisecondDurationToWindowPeriod()` get used? It makes sure that the default windowPeriod per timeRange, is same in both of these cases:

* Scenario 1:  DatePicker has -15m
   ```
   sample.data(set: "airSensor")
        |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
   ```
* Scenario 2:  user writes range for -15m
   ```
   sample.data(set: "airSensor")
        |> range(start: -15m, stop: now())
   ```

kristinarobinson replied:
@asalem1 - Here the logic is return a preset windowPeriod if the time range is non-custom, OR if custom range is used, then calculate the windowPeriod (see calcWindowPeriodForRangeDuration below).  The function below does the range / default number of points (360).

wiedld commented on src/shared/utils/windowPeriod/fromTimeRange.ts:
```diff
@@ -0,0 +1,44 @@
+const windowPeriodFromTimeRange = (timeRange: TimeRange): number => {
+  switch (timeRange.type) {
+    case 'selectable-duration':
+      return timeRange.windowPeriod
```
Yes. But it can still be overwritten by the user.

* Scenario 1:  user doesn't provide their own v.windowPeriod. So use the windowPeriod for that timeRange. In this case, `SELECTABLE_TIME_RANGES` says it's 10000.
   ```
   sample.data(set: "airSensor")
        |> range(start: -5m, stop: now())
   ```
* Scenario 2:  user overwrites the v.windowPeriod. Take the user's input, and set to 2444.
   ```
   v = {windowPeriod: 2444}
   sample.data(set: "airSensor")
        |> range(start: -5m, stop: now())
   ```

wiedld commented on src/shared/utils/ast/visitors.ts:
```diff
@@ -0,0 +1,388 @@
+    return acc
+  }
+
+  switch (node.type) {
```
This is one of the classic approaches to walking an AST tree. AST is a tree of nodes, with each node having a type. Know how to handle each node type (independent of the rest of the tree), and then recursively call the immediate children of this node.

wiedld commented on src/shared/utils/windowPeriod/fromTimeRange.ts:
```diff
@@ -0,0 +1,44 @@
+    return millisecondsToDuration(windowPeriodFromTimeRange(timeRange))
+  } catch (e) {
+    console.warn(e)
+    return null
```
Excellent point! tl;dr -- that gets decided in the next followup PR.

This method `getWindowPeriodDurationFromTimeRange()` is not yet used in the UI application. It's provided as part of the module that allows calculation of windowPeriod from various sources:
* timeRange only
  * file `./fromTimeRange.ts`
* flux query + timeRange UI variables
  * file `./fromAST.ts`
* from a numerical duration (a.k.a. timeRange length in milliseconds). 
  * file `./fromDuration.ts`

When we go to use this method, then we may decide to have it always return null so we explicitly know if failed (then we can use the fallback windowPeriod convert to a duration string). Or we may provide the fallback duration automatically. I think it comes down to (in the next PR), of when/where to handle the fallback.

wiedld commented:
>  it's too difficult for me to parse out what is net new and what is existing functionality that has been moved around. It would be really great to get this work in, however, given the PR size, the scope of the changes, and the fact that it's a long weekend, I'd be hesitant to stamp this in...

I agree that it may take more time to review. The intent is not to merge today. I added @rockstar as a reviewer, since I believe he would be the most interested party, and he is out until Tuesday.

Having said that, here is a summary of how these changes are not yet touching the UI application. The only thing consumed by the rest of the UI is the legacy `findNodes()` function, which I left unchanged.

Everything in `src/shared/utils/duration` is existing, unchanged, and is used in the UI.

Versus the `src/shared/utils/ast` and `src/shared/utils/windowPeriod`, which are new but not used:

<img width="250" alt="Screen Shot 2022-07-01 at 2 09 26 PM" src="https://user-images.githubusercontent.com/10232835/176968439-7cf3de6a-a900-49c3-9504-720b1d1f9065.png">

<img width="250" alt="Screen Shot 2022-07-01 at 2 09 42 PM" src="https://user-images.githubusercontent.com/10232835/176968448-f5c6c613-f5f7-42e5-88bc-eb9539abb5e7.png">

wiedld commented:
Note: I also removed the breakup of the `src/shared/utils/duration.ts` into 2 files. Since that was all unchanged code anyways...and is now easier for reviewers. Updated PR diff (minus the test fixtures) is `+1,246 -38`.

wiedld commented:
Based on feedback, have broken this PR into 9 separate PRs.
